src/clients/c/tb_client.zig

```diff
@@ -49,6 +53,77 @@ const TestingContext = blk: {
     break :blk ContextType(EchoClient);
 };
 
+/// Logging is global per application; it would be nice to be able to define a different logger for
+/// each client instance, though.
+var logging: Logging = .{};
+
+pub const Logging = struct {
+    const Callback = *const fn (
+        message_level: u8,
+        message_ptr: [*]const u8,
+        message_len: usize,
+    ) callconv(.C) void;
+
+    const log_line_max = 8192;
+
+    /// A logger which defers to an application provided handler.
+    pub fn application_logger(
+        comptime message_level: std.log.Level,
+        comptime scope: @Type(.EnumLiteral),
+        comptime format: []const u8,
+        args: anytype,
+    ) void {
+        // Messages are silently dropped if no logging callback is specified - unless they're warn
+        // or err. The value in having those for debugging is too high to silence them, even until
+        // client libraries catch up and implement a callback handler.
+        const callback = logging.callback orelse {
+            if (message_level == .warn or message_level == .err) {
```

Comment:
I think we could let our `root` file specify the verbosity in all cases (so it's consistent with or without a custom handler).

After all, we do want to use `warn` only for cases that the operator must be aware of, so it should not be too verbose.

Reply:
> I think we could let our root file specify the verbosity in all cases (so it's consistent with or without a custom handler).

Hmm - the downside is that it's then impossible to control the level at runtime, which is what we want most of the time. 

I actually have a WIP PR that does the opposite entirely - for replicas (and, it would be the same here) - making our default log level `.debug` and filtering at runtime (performance hit is negligible!) so that you can do something like `--log-debug` on the production binary, which is very useful.

Once this PR lands though and clients use it, I'd remove this custom handler when no callback is set entirely (or, make all messages go to stderr with a warning to fix your client). The client would potentially be able to enable `.debug` logs with an extra argument to `register_log_callback` (because the FFI call overhead could stack up) but if they wanted to filter .info or higher, that should be done with their logging library.

The only reason for this specific check for now is because we're being noisy on `stderr` which we really shouldn't be, so it's trying to minimize that.